File: dsa/queues/elixir/stack.ex
defmodule Stack do
  def new, do: []

  # O(1)
  def push(stack, element), do: [element | stack]

  # O(1)
  def pop([]), do: {nil, []}
  def pop([head | tail]), do: {head, tail}
end

stack = Stack.new

stack = Stack.push(stack, 1)
stack = Stack.push(stack, 2)
stack = Stack.push(stack, 3)

{element, stack} = Stack.pop(stack)
IO.inspect(element)

{element, stack} = Stack.pop(stack)
IO.inspect(element)

{element, _} = Stack.pop(stack)
IO.inspect(element)

File: dsa/queues/golang/stack.go
package main

import "fmt"

func push(stack []int, element int) []int {
  stack = append(stack, element)
  return stack
}

func pop(stack []int) (*int, []int) {
  if len(stack) == 0 {
    return nil, stack
  }

  element := stack[len(stack) - 1]
  tail := stack[:len(stack) - 1]

  return &element, tail
}

func main() {
  var stack []int

  stack = push(stack, 1)
  stack = push(stack, 2)
  stack = push(stack, 3)

  for {
    if element, newStack := pop(stack); element != nil {
      fmt.Println(*element)
      stack = newStack
    } else {
      break
    }
  }

  stack = push(stack, 4)

  element, _ := pop(stack)
  fmt.Println(*element)
}
